dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/EndpointErrorDecoder.java

```diff
@@ -0,0 +1,268 @@
+            case 503:
+                return Optional.of(QosException.unavailable(qosReason(response)));
+        }
+        return Optional.empty();
+    }
+
+    private @Nullable String extractErrorName(byte[] body) {
+        try {
+            NamedError namedError = NAMED_ERROR_DESERIALIZER.deserialize(new ByteArrayInputStream(body));
+            if (namedError == null) {
+                return null;
+            }
+            return namedError.errorName();
+        } catch (IOException | RuntimeException e) {
+            return null;
+        }
+    }
+
+    private T decodeInternal(Response response) {
+        Optional<RuntimeException> maybeQosException = checkCode(response);
+        if (maybeQosException.isPresent()) {
+            throw maybeQosException.get();
+        }
+        int code = response.code();
+
+        byte[] body;
+        try {
+            body = toByteArray(response.body());
+        } catch (NullPointerException | IOException e) {
+            UnknownRemoteException exception = new UnknownRemoteException(code, "<unparseable>");
+            exception.initCause(e);
+            throw exception;
+        }
+
+        Optional<String> contentType = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
+        if (contentType.isPresent()
+                && Encodings.matchesContentType(JSON_ENCODING.getContentType(), contentType.get())) {
+            try {
+                String errorName = extractErrorName(body);
+                if (errorName == null) {
+                    throw createRemoteException(body, code);
+                }
+                Deserializer<? extends T> deserializer = errorNameToJsonDeserializerMap.get(errorName);
+                if (deserializer == null) {
+                    throw createRemoteException(body, code);
+                }
+                return deserializer.deserialize(new ByteArrayInputStream(body));
+            } catch (RemoteException remoteException) {
+                // rethrow the created remote exception
+                throw remoteException;
+            } catch (Exception e) {
+                throw new UnknownRemoteException(code, new String(body, StandardCharsets.UTF_8));
+            }
+        }
+
+        throw new UnknownRemoteException(code, new String(body, StandardCharsets.UTF_8));
+    }
+
+    private static RemoteException createRemoteException(byte[] body, int code) throws IOException {
+        SerializableError serializableError = JSON_ENCODING
+                .deserializer(new TypeMarker<SerializableError>() {})
+                .deserialize(new ByteArrayInputStream(body));
+        return new RemoteException(serializableError, code);
+    }
+
+    private static byte[] toByteArray(InputStream body) throws IOException {
+        try (Reader reader = new InputStreamReader(body, StandardCharsets.UTF_8)) {
+            return CharStreams.toString(reader).getBytes(StandardCharsets.UTF_8);
+        }
```

Comment:
We don't need to convert the bytestream into a charstream (reader) to then convert back again in the other direction!

```suggestion
        try (body) {
            return body.readAllBytes();
        }
```

Reply:
🤦 This was dumb. Fixed.